diff --git a/include/linux/sched.h b/include/linux/sched.h
index 467d26046416..268cde0b4fad 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -1280,6 +1280,7 @@ struct task_struct {
 	unsigned long			lowest_stack;
 	unsigned long			prev_lowest_stack;
 #endif
+	unsigned long			sync_counter;	// Sorouri 2
 
 	/*
 	 * New fields for task_struct should be added above here, so that
diff --git a/kernel/sched/rt.c b/kernel/sched/rt.c
index e591d40fd645..69334e79ad52 100644
--- a/kernel/sched/rt.c
+++ b/kernel/sched/rt.c
@@ -1575,6 +1575,9 @@ static struct task_struct *pick_next_task_rt(struct rq *rq)
 		return NULL;
 
 	p = _pick_next_task_rt(rq);
+	// sorouri
+	printk("HHHHHHHHHHHHHHHHHHHHHHHH scheduling process pid=%u, sync_count=%lu", p->pid, p->sync_counter);
+	// sorouri
 	set_next_task_rt(rq, p, true);
 	return p;
 }
diff --git a/mm/process_vm_access.c b/mm/process_vm_access.c
index 357aa7bef6c0..694cb0aa9a14 100644
--- a/mm/process_vm_access.c
+++ b/mm/process_vm_access.c
@@ -265,10 +265,27 @@ static ssize_t process_vm_rw(pid_t pid,
 	struct iov_iter iter;
 	ssize_t rc;
 	int dir = vm_write ? WRITE : READ;
+	struct task_struct *target_task = NULL;
 
 	if (flags != 0)
 		return -EINVAL;
 
+	// sorouri
+
+	if (current->pid == pid)
+	{
+		if (dir == READ)
+		{
+			printk("a reader has jut requested to block-free wait: %u %lu\n", pid, current->sync_counter);
+			current->sync_counter += liovcnt;
+			// maye adding some scheduling codes to release the CPU.
+			return 0; //rc = 0;
+			//goto free_iovecs;
+		}
+	}
+
+	// sorouri
+
 	/* Check iovecs */
 	rc = import_iovec(dir, lvec, liovcnt, UIO_FASTIOV, &iov_l, &iter);
 	if (rc < 0)
@@ -283,6 +300,15 @@ static ssize_t process_vm_rw(pid_t pid,
 
 	rc = process_vm_rw_core(pid, &iter, iov_r, riovcnt, flags, vm_write);
 
+	// change current_task_called to the target
+	if (dir == WRITE)
+	{
+		//current_task_called->sync_counter = current_task_called->sync_counter < riovcnt ? 0 : current_task_called->sync_counter - riovcnt;
+		printk("a writer has jut requested to block-free wait: %u %lu\n", pid, riovcnt);
+		target_task = find_task_by_vpid(pid);
+		target_task->sync_counter = target_task->sync_counter > riovcnt ? target_task->sync_counter - riovcnt : 0;
+	}
+
 free_iovecs:
 	if (iov_r != iovstack_r)
 		kfree(iov_r);
